File: LeetCode/Medium/0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp;
        dp.push_back(nums[0]);
        for (int i = 1; i < nums.size(); i++) {
            if(nums[i]>dp.back()){
                dp.push_back(nums[i]);
                continue;
            }
            int j = lower_bound(dp.begin(),dp.end(),nums[i]) - dp.begin();
            dp[j] = nums[i];
        }
        return dp.size();
    }
};

File: LeetCode/Medium/0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        dp = [nums[0]]
        for i in range(1,len(nums)):
            if dp[-1] < nums[i]:
                dp.append(nums[i])
                continue
            j = bisect.bisect_left(dp, nums[i])
            dp[j] = nums[i]
        return len(dp)